service/service.go

```diff
@@ -60,25 +61,80 @@ func MakeFireAndForget(client Client) Client {
 // Server is an abstraction representing each server made available to receive inbound connections.
 type Server interface {
 	GetID() string
+	Shutdown() error
 }
 
 // Strategy is the algorithm applied by this service when it receives requests (c.f. http://wiki.c2.com/?StrategyPattern)
 type Strategy interface {
 	Do(context.Context, *pb.TheRequest) (*pb.TheResponse, error)
 }
 
+//
+// TODO: move RequestHandler into its own file
+//
+
 // RequestHandler is a protocol-independent request/response handler interface
 type RequestHandler struct {
-	Config   *Config
-	Strategy Strategy
+	Strategy Strategy // public due to circular dependency between server and strategy
+
+	config           *Config
+	stopCh           chan struct{}
+	requestCount     int
+	counterCh        chan struct{}
+	atTerminateLimit chan bool
+}
+
+// requestCounter approximates an atomic read/write counter via channels
+func (h *RequestHandler) requestCounter() {
+	for {
+		select {
+		case <-h.counterCh:
+			h.requestCount++
+
+		// check for equality rather than >= so we we only call shutdown once
+		case h.atTerminateLimit <- h.requestCount == h.config.TerminateAfter:
+		}
+	}
+}
```

Comment:
Per in-person discussion -- I think you can just write to `stopCh` directly here after incrementing `requestCount`, rather than needing a separate `atTerminateLimit` channel.

```
// requestCounter approximates an atomic read/write counter via channels
func (h *RequestHandler) requestCounter() {
	for range h.counterCh {
		h.requestCount++
		if h.requestCount == h.config.TerminateAfter {
			log.Infof("TerminateAfter limit hit (%d), stopping [%s]", h.config.TerminateAfter, h.config.ID)
			h.stopCh <- struct{}{}
		}
	}
}
```

Reply:
This is _way_ better, thanks!